socketmode/socketmode_handler.go

```diff
@@ -0,0 +1,243 @@
+func (r *SocketmodeHandler) dispatcher(evt Event) {
+	var ishandled bool
+
+	// Some eventType can be further decomposed
+	switch evt.Type {
+	case EventTypeInteractive:
+		ishandled = r.interactionDispatcher(&evt)
+	case EventTypeEventsAPI:
+		ishandled = r.eventAPIDispatcher(&evt)
+	case EventTypeSlashCommand:
+		ishandled = r.slashCommandDispatcher(&evt)
+	default:
+		ishandled = r.socketmodeDispatcher(&evt)
+	}
+
+	if !ishandled {
+		go r.Default(&evt, r.Client)
+	}
+}
+
+// Dispatch socketmode events to the registered middleware
+func (r *SocketmodeHandler) socketmodeDispatcher(evt *Event) bool {
+	if handlers, ok := r.EventMap[evt.Type]; ok {
+		// If we registered an event
+		for _, f := range handlers {
+			go f(evt, r.Client)
+		}
+
+		return true
+	}
+
+	return false
+}
+
+// Dispatch interactions to the registered middleware
+func (r *SocketmodeHandler) interactionDispatcher(evt *Event) bool {
+	var ishandled bool = false
+
+	interaction, ok := evt.Data.(slack.InteractionCallback)
+	if !ok {
+		r.Client.log.Printf("Ignored %+v\n", evt)
+		return false
+	}
+
+	// Level 1 - socketmode EventType
+	ishandled = r.socketmodeDispatcher(evt)
+
+	// Level 2 - interaction EventType
+	if handlers, ok := r.InteractionEventMap[interaction.Type]; ok {
+		// If we registered an event
+		for _, f := range handlers {
+			go f(evt, r.Client)
+		}
+
+		ishandled = true
+	}
+
+	// Level 3 - interaction with actionID
+	blockActions := interaction.ActionCallback.BlockActions
+	// outmoded approach won`t be implemented
+	// attachments_actions := interaction.ActionCallback.AttachmentActions
+
+	for _, action := range blockActions {
+		if handler, ok := r.InteractionBlockActionEventMap[action.ActionID]; ok {
+
+			go handler(evt, r.Client)
+
+			ishandled = true
+		}
+	}
+	return ishandled
+}
+
+// Dispatch eventAPI events to the registered middleware
+func (r *SocketmodeHandler) eventAPIDispatcher(evt *Event) bool {
+	var ishandled bool = false
+	eventsAPIEvent, ok := evt.Data.(slackevents.EventsAPIEvent)
+	if !ok {
+		r.Client.log.Printf("Ignored %+v\n", evt)
+		return false
+	}
+
+	innerEventType := slackevents.EventsAPIType(eventsAPIEvent.InnerEvent.Type)
+
+	// Level 1 - socketmode EventType
+	ishandled = r.socketmodeDispatcher(evt)
+
+	// Level 2 - EventAPI EventType
+	if handlers, ok := r.EventApiMap[innerEventType]; ok {
+		// If we registered an event
+		for _, f := range handlers {
+			go f(evt, r.Client)
+		}
+
+		ishandled = true
+	}
+
+	return ishandled
+}
+
+// Dispatch SlashCommands events to the registered middleware
+func (r *SocketmodeHandler) slashCommandDispatcher(evt *Event) bool {
+	var ishandled bool = false
+	slashCommandEvent, ok := evt.Data.(slack.SlashCommand)
+	if !ok {
+		r.Client.log.Printf("Ignored %+v\n", evt)
+		return false
+	}
+
+	// Level 1 - socketmode EventType
+	ishandled = r.socketmodeDispatcher(evt)
+
+	// Level 2 - SlackCommand by name
+	if handler, ok := r.SlashCommandMap[slashCommandEvent.Command]; ok {
```

Comment:
Since the slash command is tied to the actual command string (like `/command`), I think that it should not be handled by the event type. (Or if you really want to handle them, you should process level 2 before level 1.)

Reply:
I think that If you want to apply middleware functioins to all slash commands, you might want to wrap each handler functions by middleware functions.

Reply:
The order doesn't matter since each dispatcher starts a goroutine for each matching event. One event can be handler both by level 1 and level 2 if they exist.

Ideally, the level 1 should not be used. But some interesting events like `shortcuts` that work the same way as slash commands are not yet implemented. Therefore, to filter shortcuts by name you can create a middleware function that patches the missing feature.

In a way, it is possible to write all the dispatcher logic in a recursive way, When you want to handle a slash command you register a middleware that captures that specific command. In this scenario, one would only have `socketmodeDispatcher`. The issue with this option is that if you have two commands then you have two middleware functions and both would be called when a `slash_command` event comes in. But only one will actually do something.

Reply:
> The order doesn't matter since each dispatcher starts a goroutine for each matching event. One event can be handler both by level 1 and level 2 if they exist.

Sorry, this is my misunderstanding. 

> But some interesting events like shortcuts that work the same way as slash commands are not yet implemented. Therefore, to filter shortcuts by name you can create a middleware function that patches the missing feature.

I'd rather avoid misuse than deal with unimplemented features such as shortcuts.

> In a way, it is possible to write all the dispatcher logic in a recursive way, When you want to handle a slash command you register a middleware that captures that specific command. In this scenario, one would only have socketmodeDispatcher. The issue with this option is that if you have two commands then you have two middleware functions and both would be called when a slash_command event comes in. But only one will actually do something.

Sorry, I couldn't understand what you mean. Please tell me a little more detail.

Reply:
Yes, instead of using Maps and dispatchers, like this:

```go
// Register a middleware or handler for an Event (from slackevents)
func (r *SocketmodeHandler) HandleEvents(et slackevents.EventsAPIType, f SocketmodeHandlerFunc) {
	r.EventApiMap[et] = append(r.EventApiMap[et], f)
}
```

You can achieve the same goal with middleware function

```go
// Register a middleware or handler for an Event (from slackevents)
func (r *SocketmodeHandler) HandleEvents(et slackevents.EventsAPIType, f SocketmodeHandlerFunc) {
	r.Handle(EventTypeInteractive, func(e *Event, c *Client){
             interaction, _ := evt.Data.(slack.InteractionCallback)
             if interaction.type == et {
                 f(e, c)
             }
        })
}
```

I thought that what you meant by

> I think that If you want to apply middleware functioins to all slash commands, you might want to wrap each handler functions by middleware functions.

Reply:
@xNok 

Sorry for slow response.

As mentioned above, in most cases, Level 3 works as intended for the library user,
but as you approach Level 1, the possibility of unintended behavior is likely to increase.

The official library "bolt" also does not allow registering multiple handlers to for a single event(or action_id, slash command, ...),
so it can be confusing for people using other languages like Java, Python.

ref: https://github.com/slackapi/java-slack-sdk/blob/main/bolt/src/main/java/com/slack/api/bolt/App.java

I think it would be easier to use something closer to the bolt API, but... What do you think?

Reply:
@xNoK What do you think??

Reply:
This is a good point I did not look into the bold library. 

I will refactor to make it closer to the bolt API

By any chance do you know if bolt library implement the same spec in JAVA and Python?

Reply:
I have to admit that I like the JAVA implementation that separates the handler registration in `App` and `SocketModeApp` that is in charge of collecting the events. 

```
  public static void main(String[] args) throws Exception {
    String botToken = System.getenv("SLACK_BOT_TOKEN");
    App app = new App(AppConfig.builder().singleTeamBotToken(botToken).build());
    String appToken = System.getenv("SLACK_APP_TOKEN");
    SocketModeApp socketModeApp = new SocketModeApp(appToken, app);
    socketModeApp.start();
  }
```

Reply:
> By any chance do you know if bolt library implement the same spec in JAVA and Python?

I'm not very confident so I'll take a look at the implementation.

Reply:
maybe make sense merge this pr as experimental and allow use middleware. In parallel thinking about more perfect solution like bolt?

Reply:
@xNok Is there anything I can do to help?